helpers.c

```diff
@@ -26,9 +29,68 @@ int sync_file_range(int fd, uint64_t offset, uint64_t nbytes,
 }
 #endif
 
+#ifdef CONFIG_MAC_FADVISE_DISCARD
+/**
+ * CONFIG_MAC_FADVISE_DISCARD should not be run within Rosetta. Native arm64 or x86 only.
+ */
+
+#define MMAP_CHUNK_SIZE		(1024 * 1024 * 1024)
+
+int discard_pages(int fd, off_t offset, off_t size)
+{
+	caddr_t *addr;
+	uint64_t chunk_size = MMAP_CHUNK_SIZE;
+
+	printf("fsync'ing file, then invalidating UBC.\n");
+	if (fsync(fd) < 0) {
+		fprintf(stderr, "Cannot fsync file\n");
+		exit(1);
+	}
+
+	/*
+	 * mmap the file in 1GB chunks and msync(MS_INVALIDATE).
+	 */
+	while (size > 0) {
+		uint64_t mmap_size = MIN(chunk_size, size);
+
+		addr = mmap((caddr_t)0, mmap_size, PROT_NONE, MAP_SHARED, fd, offset);
+		if (addr == MAP_FAILED) {
+			fprintf(stderr, "Failed to mmap (%s), offset = %llu, size = %llu\n",
+				strerror(errno), offset, mmap_size);
+			return -1;
```

Comment:
You should return errno for the failure cases, not -1. For this one and other spots in this function.